base/iterator.jl

```diff
@@ -355,13 +380,17 @@ end
 @inline next(p::Prod2, st) = prod_next(p, st)
 @inline done(p::AbstractProdIterator, st) = st[4]
 
+# n iterators
 immutable Prod{I1, I2<:AbstractProdIterator} <: AbstractProdIterator
     a::I1
     b::I2
 end
-
 product(a, b, c...) = Prod(a, product(b, c...))
+
 eltype{I1,I2}(::Type{Prod{I1,I2}}) = tuple_type_cons(eltype(I1), eltype(I2))
+size(p::Prod) = (length(p.a), size(p.b)...)
```

Comment:
mmm, not sure I follow. Can you write a test here that would make the current implementation fail?